src/Symfony/Component/DependencyInjection/Loader/Configurator/ContainerConfigurator.php

```diff
@@ -123,7 +123,13 @@ function inline_service(string $class = null): InlineServiceConfigurator
  */
 function service_locator(array $values): ServiceLocatorArgument
 {
-    return new ServiceLocatorArgument(AbstractConfigurator::processValue($values, true));
+    $values = AbstractConfigurator::processValue($values, true);
+
+    if (isset($values[0])) {
+        trigger_deprecation('symfony/dependency-injection', '6.3', 'Using numeric keys with "service_locator" argument is deprecated. The numeric keys will be replaced to the IDs of the original services in 7.0.');
```

Comment:
This will be a new feature that is not possible before 7.1. 7.0 will be 6.4 without all deprecations.

Reply:
OMG, in May 2024? :man_facepalming:

Reply:
I don't think so: this patch aims to change a default behavior by deprecating the current one, that is the feature (6.3).

Reply:
The current behavior cannot be removed before 7.0 and the deprecation notice can be added in the next minor (6.3) to announce this breaking change.

For such need, we usually add an option to opt for the new behavior. Not setting this option triggers the deprecation notice. The default behavior change in 7.0 (default value of the option) and the option is deprecated in 7.1.

Reply:
> For such need, we usually add an option to opt for the new behavior. 

Sorry but I disagree, 3-steps changes as you're describing are quite rare. Opting in the new behavior is about defining keys explicitly which will make the deprecation notice go away and make the code ready for 7.0, no need for an extra option.

Reply:
Ok I get what you're doing. So for the transition, the keys have to be defined explicitly. 👍🏻